akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala

```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
 
     private var currentRoutees: Array[ActorRef[T]] = _
 
-    override def selectRoutee(): ActorRef[T] = {
+    override def selectRoutee(msg: T): ActorRef[T] = {
       val selectedIdx = ThreadLocalRandom.current().nextInt(currentRoutees.length)
       currentRoutees(selectedIdx)
     }
+
     override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
       currentRoutees = newRoutees.toArray
     }
+  }
 
+  final class ConsistentHashingLogic[T](
+      virtualNodesFactor: Int,
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T])
+      extends RoutingLogic[T] {
+    require(system ne null, "system argument of ConsistentHashingLogic cannot be null.")
+
+    private var currentRoutees: Set[ActorRef[T]] = Set.empty
+
+    private var consistentHash: ConsistentHash[ActorRef[T]] = ConsistentHash(currentRoutees, virtualNodesFactor)
+
+    override def selectRoutee(msg: T): ActorRef[T] =
+      try {
+        mapping(msg) match {
+          case bytes: Array[Byte] => consistentHash.nodeFor(bytes)
+          case str: String        => consistentHash.nodeFor(str)
+          case x: AnyRef =>
+            val bytes = SerializationExtension(system.toClassic).serialize(x).get
+            consistentHash.nodeFor(bytes)
```

Comment:
I don't think we should mimic the untyped impl here but either have two concrete types of hash mapping interfaces one returning string and one returning bytes, or only support one of them.

If a user want's to use the serialized form by invoking the serialization infra they can do that in their mapping function.